drivers/dac/dac_ad56xx.c

```diff
@@ -0,0 +1,382 @@
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/drivers/dac.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+
+LOG_MODULE_REGISTER(dac_ad56xx, CONFIG_DAC_LOG_LEVEL);
+
+enum ad56xx_command {
+	AD56XXCOMMAND_WRITEANDUPDATECHANNEL = 3,
+	AD56XXCOMMAND_SOFTWARERESET = 6,
+};
+
+struct ad56xx_config {
+	struct spi_dt_spec bus;
+	const struct gpio_dt_spec gpio_reset;
+	uint8_t resolution;
+	uint32_t reset_low_time_in_ns;
+	uint32_t reset_wait_time_in_ns;
+	uint8_t *channel_addresses;
+	size_t channel_count;
+};
+
+struct ad56xx_data {
+};
+
+static int ad56xx_write_command(const struct device *dev, enum ad56xx_command command,
+				uint8_t address, uint16_t value)
+{
+	const struct ad56xx_config *config = dev->config;
+	uint8_t buffer_tx[3];
+	uint8_t buffer_rx[ARRAY_SIZE(buffer_tx)];
+	const struct spi_buf tx_buf[] = {{
+		.buf = buffer_tx,
+		.len = ARRAY_SIZE(buffer_tx),
+	}};
+	const struct spi_buf rx_buf[] = {{
+		.buf = buffer_rx,
+		.len = ARRAY_SIZE(buffer_rx),
+	}};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = ARRAY_SIZE(tx_buf),
+	};
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = ARRAY_SIZE(rx_buf),
+	};
+
+	buffer_tx[0] = (command << 4) | address;
+	value = value << (16 - config->resolution);
+	sys_put_be16(value, buffer_tx + 1);
+
+	LOG_DBG("sending to DAC %s command 0x%02X, address 0x%02X and value 0x%04X", dev->name,
+		command, address, value);
+	int result = spi_transceive_dt(&config->bus, &tx, &rx);
+
+	if (result != 0) {
+		LOG_ERR("spi_transceive failed with error %i", result);
+		return result;
+	}
+
+	return 0;
+}
+
+static int ad56xx_channel_setup(const struct device *dev, const struct dac_channel_cfg *channel_cfg)
+{
+	const struct ad56xx_config *config = dev->config;
+
+	if (channel_cfg->channel_id >= config->channel_count) {
+		LOG_ERR("invalid channel %i", channel_cfg->channel_id);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->resolution != config->resolution) {
+		LOG_ERR("invalid resolution %i", channel_cfg->resolution);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int ad56xx_write_value(const struct device *dev, uint8_t channel, uint32_t value)
+{
+	const struct ad56xx_config *config = dev->config;
+
+	if (value > UINT16_MAX) {
+		LOG_ERR("invalid value %i", value);
+		return -EINVAL;
+	}
+
+	if (channel > config->channel_count) {
+		LOG_ERR("invalid channel %i", channel);
+		return -EINVAL;
+	}
+
+	return ad56xx_write_command(dev, AD56XXCOMMAND_WRITEANDUPDATECHANNEL,
+				    config->channel_addresses[channel], value);
+}
+
+static void ad56xx_busy_wait_ns(uint32_t time)
+{
+	time = time / 1000;
+
+	/*
+	 * These drivers usually require less than 1us for the reset pulse and wait.
+	 * Unfortunately, we can only wait with us precision and we should at least
+	 * wait a certain minimum amount of time, hence the fallback to 1us.
+	 */
+	if (time == 0) {
+		time = 1;
+	}
+
+	k_busy_wait(time);
+}
```

Comment:
Thought about the same thing, but I also did want to copy the exact values from the datasheet into the driver implementation. But maybe it would also be sufficient to just move the comment to the define of the value and set it to 1ms?

Reply:
1us, of course

Reply:
done